nomt/src/seglog/mod.rs

```diff
@@ -0,0 +1,1267 @@
+                    last.next(),
+                );
+            }
+
+            let (was_active, became_active, became_inactive) =
+                self.on_next_record(index, record_id);
+
+            if was_active || became_active || became_inactive {
+                seg_reader.read_payload(&mut self.payload_buf)?;
+                process_record(record_id, &self.payload_buf)?;
+            } else {
+                seg_reader.skip_payload()?;
+            }
+
+            // Update the minimum and maximum record IDs observed in the segment.
+            if min.is_none() {
+                min = Some(record_id);
+            }
+            if max.map_or(true, |max| record_id > max) {
+                max = Some(record_id);
+            }
+            last = Some(record_id);
+        }
+        // If the segment is empty, we set the min and max to nil. They are always Some after this
+        // function returns.
+        let candidate = &mut self.candidates[index];
+        candidate.min = min.unwrap_or(0.into()).into();
+        candidate.max = max.unwrap_or(0.into()).into();
+        Ok(())
+    }
+
+    fn on_next_record(&mut self, segment_index: usize, record_id: RecordId) -> (bool, bool, bool) {
+        let was_active = self.is_active();
+        let became_active = self.enter_active(segment_index, record_id);
+        let became_inactive = self.exit_active(segment_index, record_id);
+        (was_active, became_active, became_inactive)
+    }
+
+    fn enter_active(&mut self, segment_index: usize, record_id: RecordId) -> bool {
+        if self.start_active.is_nil() {
+            return false;
+        }
+        if self.active_segment_start.is_none() && record_id >= self.start_active {
+            self.active_segment_start = Some(segment_index);
+            return true;
+        }
+        false
+    }
+
+    fn exit_active(&mut self, segment_index: usize, record_id: RecordId) -> bool {
+        if self.start_active.is_nil() {
+            return false;
+        }
+        if self.is_active() && self.active_segment_end.is_none() && record_id >= self.end_active {
+            self.active_segment_end = Some(segment_index);
+            return true;
+        }
+        false
+    }
+
+    /// Returns `true` if the given record ID is considered active.
+    fn is_active(&self) -> bool {
+        // If the `start_active` is nil, then the record trivially doesn't belong to the active
+        // range.
+        if self.start_active.is_nil() {
+            assert!(self.end_active.is_nil());
+            return false;
+        }
+        self.active_segment_start.is_some() && self.active_segment_end.is_none()
+    }
+
+    // /// Find the index of the first and last active segments.
+    // ///
+    // /// An active segment is a segment that contains records in the active range.
+    // ///
+    // /// Also makes some checks on the segments.
+    // fn find_active_segments(&mut self) -> Result<RangeInclusive<usize>> {
```

Comment:
is this meant to be committed? will it be reinstated in a follow up?

Reply:
Meant to be removed.

Reply:
Check v3.